backends/mongodb/mongodb.c

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#define BACKENDS_INTERNALS
+#include "mongodb.h"
+#include <mongoc.h>
+
+#define CONFIG_FILE_LINE_MAX ((CONFIG_MAX_NAME + CONFIG_MAX_VALUE + 1024) * 2)
+
+mongoc_client_t *mongodb_client;
+mongoc_collection_t *mongodb_collection;
+
+int mongodb_init(const char *uri_string, const char *database_string, const char *collection_string, int32_t socket_timeout) {
+    mongoc_uri_t *uri;
+    bson_error_t error;
+
+    mongoc_init();
+
+    uri = mongoc_uri_new_with_error(uri_string, &error);
+    if(unlikely(!uri)) {
+        error("BACKEND: failed to parse URI: %s. Error message: %s", uri_string, error.message);
+        return 1;
+    }
+
+    socket_timeout = mongoc_uri_get_option_as_int32(uri, MONGOC_URI_SOCKETTIMEOUTMS, socket_timeout);
```

Comment:
Yes, it is a fallback value. I think we don't need to rename the variable because 99,99% of the time it will be the actual value of the socket timeout.

Reply:
The problem is that this is not reflected in the naming (both variable and configuration) and when someone sets the timeout value to some value, they will expect it to be applied. When it isn't, they will — very justifiably — be upset. 

So, in my opinion the variable should, at the very least, be clear that it's a fallback value. But there's a more fundamental 'design' issue here: typically, passed parameters are _overriding_ pre-existing values, not used as fallbacks. 

Even though it's a relatively minor point from a technical standpoint, it is quite important in terms of user experience: the implementation, as it is today, can be confusing, both to developers and end-users.

Reply:
> The problem is that this is not reflected in the naming (both variable and configuration) 

Hm... I don't understand. What "configuration" are you talking about? The variable which is passed to `mongoc_uri_get_option_as_int32()` is internal and does not represent any configuration option.

> when someone sets the timeout value to some value, they will expect it to be applied

I will be applied. If the user sets the socket timeout using MongoDB URI, it overrides the internal variable.

> But there's a more fundamental 'design' issue here: typically, passed parameters are overriding pre-existing values, not used as fallbacks.

It's the design of the `libmongoc` library. We also have the same design for the `config_get` and `config_set` functions in Netdata.

> the implementation, as it is today, can be confusing, both to developers and end-users

I see that the variable name can be confusing to developers. I'll change its name. But how it could be confusing to end-users?

Reply:
> It's the design of the libmongoc library. We also have the same design for the config_get and config_set functions in Netdata.

Not at all. `libmongoc` doesn't require you to "get and set" a value in the way it's done. What you're doing is _correct_, and your thinking is absolutely valid (i.e. allow the user to set a timeout in the connection URI, then, failing that, use a sane default), but — without documentation — the implementation can confuse people as it is.

Plus we might have wanted to expose the timeout in the configuration file at some point, instead of calculating using the `global_backend_update_every`.

Anyway, I think renaming the variable works for me as this is turning into more of a theoretical discussion about design and conventions than a CR :)